cuqi/experimental/mcmc/_hmc.py

```diff
@@ -89,13 +89,16 @@ class NUTS(Sampler):
 
     _STATE_KEYS = Sampler._STATE_KEYS.union({'_epsilon', '_epsilon_bar',
                                                 '_H_bar', '_mu',
-                                                '_alpha', '_n_alpha'})
+                                                '_alpha', '_n_alpha',
+                                                'current_target_logd',
+                                                'current_target_grad',
+                                                'max_depth'})
```

Comment:
I feel like max_depth is not a state? I would be similar to max_it in linearRTO

Reply:
Good point.

The checkpointing test below fails if I removed max depth. In checkpoint targets list below I added NUTS with specific tree depth. , The `sampler_fresh`, loses this information if max_depth is not part of the state. This is possibly up to us if we want such information to be part of the state or not. Do we want the sampler that loads the checkpoint to continue exactly as the original sampler?

```python
# List of all samplers from cuqi.experimental.mcmc that should be tested for checkpointing + their parameters
checkpoint_targets = [
    cuqi.experimental.mcmc.ULA(cuqi.testproblem.Deconvolution1D().posterior, scale=0.0001),
    cuqi.experimental.mcmc.MALA(cuqi.testproblem.Deconvolution1D().posterior, scale=0.0001),
    cuqi.experimental.mcmc.LinearRTO(cuqi.testproblem.Deconvolution1D().posterior),
    cuqi.experimental.mcmc.UGLA(create_lmrf_prior_target(dim=16)),
    cuqi.experimental.mcmc.Direct(cuqi.distribution.Gaussian(np.zeros(10), 1)),
    cuqi.experimental.mcmc.Conjugate(create_conjugate_target("Gaussian-Gamma")),
    cuqi.experimental.mcmc.ConjugateApprox(create_conjugate_target("LMRF-Gamma")),
    cuqi.experimental.mcmc.NUTS(cuqi.testproblem.Deconvolution1D(dim=10).posterior, max_depth=4)
]
    
# List of samplers from cuqi.experimental.mcmc that should be skipped for checkpoint testing


@pytest.mark.parametrize("sampler", checkpoint_targets)
def test_checkpointing(sampler: cuqi.experimental.mcmc.Sampler):
    """ Check that the checkpointing functionality works. Tested with save_checkpoint(filename) and load_checkpoint(filename).
    This also implicitly tests the get_state(), set_state(), get_history(), and set_history() as well as the reset() methods.
    
    """

    # Run sampler with some samples
    sampler.warmup(50).sample(50)

    # Save checkpoint
    sampler.save_checkpoint('checkpoint.pickle')

    # Reset the sampler, e.g. remove all samples but keep the state
    # Calling sampler.reset() would fail since it resets state.
    # Instead, we want to call a method like reset_history().
    # Currently we do this:
    sampler._samples = []

    # Do some more samples from pre-defined rng state
    np.random.seed(0)
    samples1 = sampler.warmup(50).sample(50).get_samples().samples

    #TODO Consider changing now that target=None is allowed.
    # Now load the checkpoint on completely fresh sampler not even with target
    sampler_fresh = sampler.__class__(sampler.target) # In principle init with no arguments. Now still with target
    sampler_fresh.load_checkpoint('checkpoint.pickle')

    # Do some more samples from pre-defined rng state
    np.random.seed(0)
    samples2 = sampler_fresh.warmup(50).sample(50).get_samples().samples

    # Check that the samples are the same
    assert np.allclose(samples1, samples2), f"Samples1: {samples1}\nSamples2: {samples2}"

```

Reply:
We discussed and agree to to max_depth in state for now